content/blog/2022-05-19-fast-brili.md

```diff
@@ -0,0 +1,141 @@
+require multiple words. This means that our array index for iteration will
+NOT be an induction variable, so the C compiler will not be able to optimize it
+as aggressively. One solution to this could be using a less flat representation 
+of a list of instructions, and using pointers for the multiword 
+instructions, but this seems like more trouble than it's worth, especially 
+considering the benefits of cache locality.
+
+For the armv8 assembly generation, we initially implemented it as a single pass
+over the bytecode that printed the instructions as it went. This is fine, but it
+leaves little room for extensions and improvements, so in order to get started
+on register allocation we refactored to have an internal representation of arm
+instructions. For this, we went with a tagged union, as it is similar to a
+variant which is very convenient for this, as learned in CS 4120.
+
+## Evaluation
+We used the Bril benchmarks from the repo, and used clock time to evaluate the
+difference. This isn't exactly optimal as most of these benchmarks are very
+small, so the startup time might be significant, but the startup time should
+hurt our interpreter the most, as it has the translation phase. Our results are
+[here](https://github.coecis.cornell.edu/cs897/fast-bril/blob/master/eval/performance.ods),
+in ods format. On average, `brili-rs` has an 84% speedup over `brili`, and
+`fastbrili` has a 77% speedup over `brili-rs`. `fastbrili` also has a 96%
+speedup over `brili`, so I think we can call this a job well done.
+
+## Drawbacks
+While we would love to say that `fastbrili` is the answer to all of your Bril
+needs, there are some minor drawbacks, mainly that invalid Bril programs cause
+undefined behavior. Most of the type information is erased and then inferred
+(e.g. if you use `add` on two floats, it will not check, assume they are ints,
+and run integer addition on the values), but we do deal with all correct Bril
+programs correctly. We believe that this isn't much of an issue, as `brili`
+isn't actually that helpful with these errors, though we do acknowledge that
+`fastbrili` is worse.
+
+### Extensibility
+One of the really cool things about Bril is that it is relatively easy to
+extend. Our bytecode is designed in a way that it supports $2^15$ opcodes, which
+is room for plenty of extensions, but if you wanted to add more types this would
+be considerably more difficult. We rely on being able to represent types in 16
+bits, by using 2 of them for the "base type" and then the other 14 for pointer
+depth. Realistic $2^14$ is deeper than any pointer will be nested, so adding a
+few more base types wouldn't be too much of an issue, but adding more complex
+types would be a serious complication. This could be solved by having one bit of
+tag for whether the type is complex, and then using the rest of the bits to
+index into some memoized array of the types that show up, but this would require
+a bit of work.
```

Comment:
As a counterpoint to the previous section, signaling useful errors for *these* limitations seems important. (In other words, silently doing the wrong thing for programs that exceed these limits would be bad.)

Reply:
ok will add